File: genetisk/src/main/java/no/asmund/genetisk/Population.java
package no.asmund.genetisk;

import java.util.ArrayList;
import java.util.Random;

public class Population {
    Creature[] population;
    ArrayList<Creature> matingPool;
    int generations;
    int popSize;
    float mutationRate;
    float maxFit;

    public Population(float m, int num) {
        mutationRate = m;
        popSize = num;
        population = new Creature[num];
        matingPool = new ArrayList<>();
        generations = 0;

        maxFit = 0;

        for (int i = 0; i < popSize; i++) {
            population[i] = new Creature(new DNA());
        }
    }

    void fitness() {
        int max = 0;
        for (int i = 0; i < popSize; i++) {
            population[i].fitness();
            if (population[i].fitness > maxFit) {
                maxFit = population[i].fitness;
                max = i;
            }
        }
        System.out.println(population[max].fitness);
        //Normaliser
        for (int i = 0; i < popSize; i++) {
            population[i].fitness /= maxFit;
        }

    }

    void selection() {
        matingPool.clear();
        for (int i = 0; i < popSize; i++) {
            int n = (int) (population[i].fitness * 100);
            for (int j = 0; j < n; j++) {
                matingPool.add(population[i]);
            }
        }
    }

    void reproduction() {
        Random rand = new Random();
        for (int i = 0; i < popSize; i++) {
            // Velger to foreldre
            int m = rand.nextInt(matingPool.size());
            int d = rand.nextInt(matingPool.size());
            Creature mom = matingPool.get(m);
            Creature dad = matingPool.get(d);
            // Henter genene
            DNA momgenes = mom.dna;
            DNA dadgenes = dad.dna;
            // Parrer moren og faren
            DNA child = momgenes.crossover(dadgenes);
            child.mutation();
            population[i] = new Creature(child);
        }
    }
    /*    public void live() {
        for (int i = 0; i < popSize; i++) {
            population[i].run();
        }
    }*/
}

File: genetisk/src/main/java/no/asmund/genetisk/Population.kt
package no.asmund.genetisk

import kotlin.random.Random

class Population(val mutationRate: Float, val popSize: Int) {
    val population: Array<Creature>
    val matingPool: MutableList<Creature> = mutableListOf()
    var generations: Int = 0
    var maxFit: Float = 0f

    init {
        population = Array(popSize) { Creature(DNA()) }
    }

    fun fitness() {
        var max = 0
        for (i in 0 until popSize) {
            population[i].fitness()
            if (population[i].fitness > maxFit) {
                maxFit = population[i].fitness
                max = i
            }
        }
        println(population[max].fitness)

        // Normalize fitness values
        for (i in 0 until popSize) {
            population[i].fitness /= maxFit
        }
    }

    fun selection() {
        matingPool.clear()
        for (i in 0 until popSize) {
            val n = (population[i].fitness * 100).toInt()
            repeat(n) {
                matingPool.add(population[i])
            }
        }
    }

    fun reproduction() {
        for (i in 0 until popSize) {
            // Select two parents
            val m = Random.nextInt(matingPool.size)
            val d = Random.nextInt(matingPool.size)
            val mom = matingPool[m]
            val dad = matingPool[d]

            // Get their genes
            val momGenes = mom.dna
            val dadGenes = dad.dna

            // Create child by crossing over parents
            val child = momGenes.crossover(dadGenes)
            child.mutation(mutationRate)
            population[i] = Creature(child)
        }
        generations++
    }
}